src/firmware_new/boards/frankie_v1/io/ublox_odinw262_communicator.c

```diff
@@ -0,0 +1,209 @@
+#include "firmware_new/boards/frankie_v1/usart.h"
+#include "main.h"
+#include "task.h"
+
+// TODO have a way of configuring these values externally
+// https://github.com/UBC-Thunderbots/Software/issues/1876
+#define WIFI_SSID "thunderbots"
+#define WIFI_PASS "tbots1234"
+
+#define DMA_BUFFER __attribute__((section(".dma_buffer")))
+#define RX_BUFFER_LENGTH_BYTES 4096
+
+// Buffers: The DMA_BUFFER is managed by the UART peripheral configured in "circular" mode
+DMA_BUFFER static uint8_t g_dma_uart_receive_buffer[RX_BUFFER_LENGTH_BYTES] = {0};
+static char g_uart_receive_buffer[RX_BUFFER_LENGTH_BYTES]                   = {0};
+
+// Interrupt state: the response status and counters are used by handleidleline
+volatile size_t g_dma_counter_on_uart_idle_line        = 0;
+volatile size_t g_last_byte_parsed_from_dma_buffer     = 0;
+volatile UbloxResponseStatus_t g_ublox_response_status = UBLOX_RESPONSE_UNDETERMINED;
+
+static osSemaphoreId_t g_dma_receive_semaphore;
+static UART_HandleTypeDef* g_ublox_uart_handle;
+static GpioPin_t* g_ublox_reset_pin;
+static uint32_t g_ublox_response_timeout;
+
+static bool g_initialized = false;
+
+void io_ublox_odinw262_communicator_init(UART_HandleTypeDef* uart_handle,
+                                         GpioPin_t* ublox_reset,
+                                         uint32_t ublox_response_timeout)
+{
+    assert(!g_initialized);
+
+    // create a binary sempahore
+    g_dma_receive_semaphore  = osSemaphoreNew(1U, 1U, NULL);
+    g_ublox_uart_handle      = uart_handle;
+    g_ublox_reset_pin        = ublox_reset;
+    g_ublox_response_timeout = ublox_response_timeout;
+
+    if (HAL_UART_Init(g_ublox_uart_handle) != HAL_OK)
+    {
+        TLOG_FATAL("Failed to initialize UART connection");
+    }
+
+    // We use idle line detection to know when to parse the circular buffer
+    __HAL_UART_ENABLE_IT(g_ublox_uart_handle, UART_IT_IDLE);
+
+    // Setup DMA transfer to continually receive data over UART as the DMA
+    // controller is setup in circular mode for rx/tx
+    if (HAL_UART_Receive_DMA(g_ublox_uart_handle, g_dma_uart_receive_buffer,
+                             RX_BUFFER_LENGTH_BYTES) != HAL_OK)
+    {
+        TLOG_FATAL("Failed to setup UART DMA Receive Transfer");
+    }
+
+    g_initialized = true;
+}
+
+void io_ublox_odinw262_communicator_task(void* arg)
+{
+    assert(g_initialized);
+
+    io_ublox_odinw262_reset();
+    io_ublox_odinw262_communicator_waitForBoot();
+
+    // TODO handle errors and implement WiFi watchdog
+    // https://github.com/UBC-Thunderbots/Software/issues/1875
+    io_ublox_odinw262_communicator_sendATCommand("AT+UMLA=2,00AAAAAAAA00\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT&W\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+CPWROFF\r");
+    io_ublox_odinw262_communicator_waitForBoot();
+
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,0,0\r");
+    io_ublox_odinw262_communicator_sendATCommand("AT+UBRGC=0,1,3,1\r");
```

Comment:
More comments

Reply:
added a link to the section its adapted from and a link to the at command docs